controller/src/main/java/com/esm/epam/controller/CertificateController.java

```diff
@@ -25,75 +26,98 @@
+        this.hateoasBuilder = hateoasBuilder;
+    }
+
+    @GetMapping(params = {"page", "size"})
     @JsonView(View.UI.class)
-    public ResponseEntity<List<Certificate>> getCertificateList(@RequestParam(required = false) MultiValueMap<String, Object> params) throws ResourceNotFoundException, ControllerException {
-        List<Certificate> certificates = new ArrayList<>();
-        if (params.size() == 0) {
-            certificates = certificateService.getAll();
+    public ResponseEntity<List<Certificate>> getCertificateList(@RequestParam(required = false) MultiValueMap<String, Object> params, @RequestParam("page") @Min(0) int page, @RequestParam("size") @Min(1) int size) throws ResourceNotFoundException, ControllerException, ServiceException, DaoException {
+        List<Certificate> certificates;
+        if (params.size() == 2) {
+            certificates = certificateService.getAll(page, size);
         } else {
             validateSortValues(params);
-            certificates = certificateService.getFilteredList(params);
+            certificates = certificateService.getFilteredList(params, page, size);
+        }
+        for (Certificate certificate : certificates) {
+            hateoasBuilder.buildFullHateoas(certificate);
         }
         return new ResponseEntity<>(certificates, HttpStatus.OK);
     }
 
     @GetMapping("/{id}")
     @JsonView(View.UI.class)
-    public ResponseEntity<Certificate> getCertificate(@PathVariable("id") @Min(1L) Long id) throws ResourceNotFoundException, DaoException {
+    public ResponseEntity<Certificate> getCertificate(@PathVariable("id") @Min(1L) Long id) throws ResourceNotFoundException, DaoException, ControllerException, ServiceException {
         Certificate certificate = certificateService.getById(id);
+        hateoasBuilder.buildFullHateoas(certificate);
         return new ResponseEntity<>(certificate, HttpStatus.OK);
     }
 
 
     @DeleteMapping("/{id}/tags/{tag_id}")
-    public ResponseEntity<Certificate> deleteCertificate(@PathVariable("id") @Min(1L) Long id, @PathVariable("tag_id") @Min(1L) Long idTag) throws ResourceNotFoundException, DaoException {
-        return certificateService.deleteTag(id, idTag)
-                .map(ResponseEntity::ok)
-                .orElseGet(() -> ResponseEntity.noContent().build());
+    @JsonView(View.UI.class)
+    public ResponseEntity<RepresentationModel<Certificate>> deleteTagCertificate(@PathVariable("id") @Min(1L) Long id, @PathVariable("tag_id") @Min(1L) Long idTag) throws DaoException, ControllerException, ServiceException, ResourceNotFoundException {
+        ResponseEntity<RepresentationModel<Certificate>> responseEntity;
+        Optional<Certificate> addedCertificate = certificateService.deleteTag(id, idTag);
+        if (addedCertificate.isPresent()) {
+            hateoasBuilder.buildFullHateoas(addedCertificate.get());
+            responseEntity = new ResponseEntity<>(addedCertificate.get(), OK);
```

Comment:
Check response HttpStatuses for your endpoints. DELETE operation based on RestApi should have 204 No content response status. Check response statuses for all other endpoints, 201 for create etc.